# VISOR

<div style="width: 180px; margin: 0 auto;"><a href="http://popmodels.cancercontrol.cancer.gov/gsr/"><img src="http://popmodels.cancercontrol.cancer.gov/gsr/static/img/gsr_tile.jpg" alt="Catalogued on GSR" width="180" height="60" /></a></div>

![alt text](VISOR.png)

## VISOR: haplotype-aware structural variants simulator for short, long and linked reads

VISOR is available to be installed from source or as a [Docker container](https://hub.docker.com/r/davidebolo1993/visor).
Please have a look at [VISOR documentation](https://davidebolo1993.github.io/visordoc/) for any installation or usage questions.

[Source Code](https://github.com/davidebolo1993/VISOR/tree/master/VISOR)

[Documentation](https://davidebolo1993.github.io/visordoc/)

## Citation

Are you using VISOR in your works? Please cite:

> Davide Bolognini, Ashley D Sanders, Jan O Korbel, Alberto Magi, Vladimir Benes, Tobias Rausch. [VISOR: a versatile haplotype-aware structural variant simulator for short and long read sequencing](https://academic.oup.com/bioinformatics/article/doi/10.1093/bioinformatics/btz719/5582674/). Bioinformatics. 2019 Oct 7.

## Related works

VISOR has been applied to simulate Strand-seq data in:

> Ashley D Sanders, Sascha Meiers, Maryam Ghareghani, David Porubsky, Hyobin Jeong, M Alexandra CC van Vliet, Tobias Rausch, Paulina Richter-Pechańska, Joachim B Kunz, Silvia Jenni, Davide Bolognini, Gabriel MC Longo, Benjamin Raeder, Venla Kinanen, Jürgen Zimmermann, Vladimir Benes, Martin Schrappe, Balca R Mardin, Andreas E Kulozik, Beat Bornhauser, Jean-Pierre Bourquin, Tobias Marschall, Jan O Korbel. [Single-cell analysis of structural variations and complex rearrangements with tri-channel processing](https://www.nature.com/articles/s41587-019-0366-x). Nature Biotechnology. 2019 Dec 23.
